File: 1.Stack/Python/StackUsingLL.py
# Node class to represent each element in the stack
class Node:
    def __init__(self, value):
        self.value = value  # The value of the node
        self.next = None    # Pointer to the next node in the stack

# Stack class implementing a stack using a linked list
class Stack:
    def __init__(self):
        self.top = None   # Reference to the top of the stack
        self.size = 0     # Number of elements in the stack

    # Method to add an element to the stack
    def push(self, value):
        node = Node(value)  # Create a new node with the given value
        node.next = self.top  # Link the new node to the current top
        self.top = node  # Update the top to the new node
        self.size += 1   # Increment the size of the stack

    # Method to remove and return the top element of the stack
    def pop(self):
        if self.top is None:  # Check if the stack is empty
            raise IndexError("Pop from empty stack")  # Raise an error if the stack is empty
        value = self.top.value  # Retrieve the value of the top node
        self.top = self.top.next  # Update the top to the next node
        self.size -= 1  # Decrement the size of the stack
        return value  # Return the popped value

    # Method to view the top element without removing it
    def peek(self):
        if self.top is None:  # Check if the stack is empty
            raise IndexError("Peek from empty stack")  # Raise an error if the stack is empty
        return self.top.value  # Return the value of the top element

    # Method to check if the stack is empty
    def is_empty(self):
        return self.top is None  # Return True if the stack is empty, False otherwise

    # Method to get the size of the stack
    def get_size(self):
        return self.size  # Return the current size of the stack

# Example Usage
stack = Stack()
stack.push(10)  # Push 10 onto the stack
stack.push(20)  # Push 20 onto the stack
print(stack.peek())  # Output: 20 (Top element of the stack)
print(stack.pop())   # Output: 20 (Remove and return the top element)
print(stack.get_size())  # Output: 1 (Size of the stack after pop)
print(stack.is_empty())  # Output: False (The stack still has elements)

File: 1.Stack/Cpp/StackUsingLL.cpp
// CPP Implementation of Stack using Linked List
#include <iostream>

// Creating Node
struct Node
{
    int data;
    Node *next;
};

// Creating Stack
struct Stack
{
    Node *top;
};
typedef struct Stack ST;

// Is Empty
int isEmpty(Stack *myStack)
{
    if (myStack->top == NULL)
        return 1;
    return 0;
}

// Pop
int pop(Stack *myStack)
{
    if (isEmpty(myStack))
    {
        std::cout << "The Stack is Empty!" << std::endl;
        return -1; // Error value
    }
    Node *temp = myStack->top;
    while (temp->next->next != NULL)
    {
        temp = temp->next;
    }

    int popped = temp->next->data;
    temp->next = NULL;
    return popped;
}

// Push
void push(Stack *myStack, int element)
{
    Node *newNode = new Node();
    newNode->data = element;
    newNode->next = myStack->top;
    myStack->top = newNode;
    std::cout << element << " pushed to the stack!" << std::endl;
}

// Peek
int peek(Stack *myStack)
{
    if (isEmpty(myStack))
    {
        std::cout << "The Stack is Empty!" << std::endl;
        return -1; // Error value
    }
    Node *temp = myStack->top;
    while (temp->next != NULL)
    {
        temp = temp->next;
    }
    return temp->data;
}

// Traverse the stack
void traverse(Stack *myStack)
{
    if (isEmpty(myStack))
    {
        std::cout << "The Stack is Empty!" << std::endl;
    }
    else
    {
        Node *temp = myStack->top;
        while (temp != NULL)
        {
            std::cout << temp->data << " ";
            temp = temp->next;
        }
        std::cout << std::endl;
    }
}

int main()
{
    // Initializing the stack
    ST *myStack = new ST();
    myStack->top = NULL;

    // Pushing elements to the stack
    push(myStack, 1);
    push(myStack, 2);
    push(myStack, 3);

    // Traversing the stack
    traverse(myStack);

    // Popping elements from the stack
    std::cout << pop(myStack) << " popped from the stack!" << std::endl;
    std::cout << pop(myStack) << " popped from the stack!" << std::endl;

    // Traversing the stack
    traverse(myStack);

    // Peeking the top element
    std::cout << "Top Element: " << peek(myStack) << std::endl;

    return 0;
}